### PR TITLE
Resolving the internal service URL for internal API calls

### PR DESCRIPTION
**Pull request by GANGANI**
### Proposed changes in this pull request
This PR changes will resolve the internal service URL for internal API calls. An optional config is introduced in “identity.xml” for the internal hostname as “<ServerHostName>”  with the default value “localhost”. It will be configurable from deployment.toml if necessary.

Relates to https://github.com/wso2/product-is/issues/8194

**Sample configurations are as follows,**
- Configuration in identity.xml.j2 file
`<ServerHostName>{{server.internal_hostname}}</ServerHostName>
`
- Configuration in deployment.toml
`[server]`
`internal_hostname = “internal.wso2.is”`

This configuration will be utilized at the service url builder utility to build the internal absolute url of a service endpoint, that will be consumed at places that will generate internal calls. As at now, we have accountrecoveryendpoint invoking internal API calls.

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java [hunk 1]
```diff
@@ -418,7 +442,7 @@ public String getRelativeInternalURL() {
             return relativeInternalUrl;
         }
 
-        private String fetchAbsoluteUrl() throws URLBuilderException {
+        private String fetchAbsolutePublicUrl() throws URLBuilderException {
```
Comment (ashensw) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:445]: Shall we change the below variable name also to `absolutePublicUrl`?

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java [hunk 2]
```diff
@@ -311,20 +311,20 @@ public void testGetAbsoluteURL(String protocol, String hostName, int port, Strin
                                 "fragment").addFragmentParameter("key2", "fragment").addFragmentParameter("key3",
                                 "fragment").addFragmentParameter("key4", "fragment").addParameter("key1", "v")
                                 .addParameter("key2", "v").addParameter("key3", "v").addParameter("key4", "v").build()
-                                .getAbsoluteInternalURL();
+                                .getAbsolutePublicURL();
```
Comment (ashensw) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java:314]: Shall we add a separate test case to cover getAbsoluteInternalURL as well?

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java (hunk 1 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:445]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2949/commits/718c68995f4548bc59738aff6b21337e7d8164ea

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java (hunk 2 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java:357]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2949/commits/2b1d2672a9098a2553b55ed732619275bc23ae27

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java [hunk 3]
```diff
@@ -236,6 +238,24 @@ private String fetchHostName() throws URLBuilderException {
         return hostName;
     }
 
+    private String fetchInternalHostName() throws URLBuilderException {
+
+        String internalHostName = IdentityUtil.getProperty(IdentityCoreConstants.SERVER_HOST_NAME);
+        try {
```
Comment (emswbandara) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:244]: here code from line 244 to 256 seems to be duplicated from code in line 226 to 238.. shall we move it to a common method?

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java [hunk 4]
```diff
@@ -304,6 +347,52 @@ public void testGetAbsoluteURL(String protocol, String hostName, int port, Strin
...
+                        ServiceURLBuilder.create().addPath(urlPath).setFragment(fragment).addParameter("key1", "v")
+                                .addParameter("key2", "v").addParameter("key3", "v").addParameter("key4", "v").build()
+                                .getAbsolutePublicURL();
+            } else {
+                absoluteUrl =
+                        ServiceURLBuilder.create().addPath(urlPath).setFragment(fragment).build().getAbsolutePublicURL();
+            }
+
+        } catch (URLBuilderException e) {
+            //Mock behaviour, hence ignored
```
Comment (emswbandara) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java:374]: use standard comment format add space after // and add fullstop at end.

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2 [hunk 5]
```diff
@@ -70,6 +70,9 @@
         <TrustManagerType>{{key_mgt.trust_manager_type}}</TrustManagerType>
     </Security>
 
+    <!--This configuration is to resolve the internal service URL for internal API calls-->
```
Comment (emswbandara) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2:73]: add space after `<!--` and before` -->`

Comment (GANGANI) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2949/commits/0c97e69333943613c0cac9c3e6dd5c2f1eab66fd

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java (hunk 4 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java]: Fixed 0c97e69

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java [hunk 6]
```diff
@@ -418,17 +442,30 @@ public String getRelativeInternalURL() {
             return relativeInternalUrl;
         }
 
-        private String fetchAbsoluteUrl() throws URLBuilderException {
+        private String fetchAbsolutePublicUrl() throws URLBuilderException {
+
+            StringBuilder absolutePublicUrl = new StringBuilder();
+            absolutePublicUrl.append(protocol).append("://");
+            absolutePublicUrl.append(hostName.toLowerCase());
```
[the remark this answers is not in the document]
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:448]: Fixed 0c97e69

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java [hunk 7]
```diff
@@ -418,17 +442,30 @@ public String getRelativeInternalURL() {
             return relativeInternalUrl;
         }
 
-        private String fetchAbsoluteUrl() throws URLBuilderException {
+        private String fetchAbsolutePublicUrl() throws URLBuilderException {
+
+            StringBuilder absolutePublicUrl = new StringBuilder();
+            absolutePublicUrl.append(protocol).append("://");
```
[the remark this answers is not in the document]
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:447]: Fixed 0c97e69

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java (hunk 3 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java]: Fixed 0c97e69

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java [hunk 8]
```diff
@@ -418,17 +435,42 @@ public String getRelativeInternalURL() {
...
+        private String fetchAbsoluteInternalUrl() throws URLBuilderException {
 
-            StringBuilder absoluteUrl = new StringBuilder();
-            absoluteUrl.append(protocol).append("://");
-            absoluteUrl.append(hostName.toLowerCase());
+            StringBuilder absoluteInternalUrl = new StringBuilder();
+            if (StringUtils.isBlank(protocol)) {
+                throw new URLBuilderException("Protocol of service URL is not available");
+            }
+            if (StringUtils.isBlank(hostName)) {
```
Comment (emswbandara) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:463]: let's add the null check for internalHostName as well.

Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2949/commits/7a8092b90d4e91354b326e60bdef0c9b34ef10cb

**Comment by emswbandara** [comment]
@GANGANI can you please check why travis is failing.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java [hunk 9]
```diff
@@ -252,29 +265,33 @@ private Integer fetchPort() {
 
         private String protocol;
         private String hostName;
```
Comment (malithie) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:267]: Shall we rename this for proxyHostName

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java [hunk 10]
```diff
@@ -80,14 +81,15 @@ public ServiceURL build() throws URLBuilderException {
 
         String protocol = fetchProtocol();
         String hostName = fetchHostName();
```
Comment (malithie) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:83]: Also this one as fetchProxyHostName

Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2949/commits/505c8f4e1f9f23a167ac4bac3b67bac5bab186cc

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java (hunk 9 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2949/commits/505c8f4e1f9f23a167ac4bac3b67bac5bab186cc

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ServiceURL.java [hunk 11]
```diff
@@ -37,7 +37,7 @@
      *
      * @return String of the host name.
      */
-    String getHostName();
+    String getProxyHostName();
```
Comment (janakamarasena) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ServiceURL.java:40]: This component did not go in a released product version yet right? We have changed the interface method name here.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java [hunk 12]
```diff
@@ -418,17 +435,42 @@ public String getRelativeInternalURL() {
             return relativeInternalUrl;
         }
 
-        private String fetchAbsoluteUrl() throws URLBuilderException {
+        private String fetchAbsolutePublicUrl() throws URLBuilderException {
+
+            StringBuilder absolutePublicUrl = new StringBuilder();
+            if (StringUtils.isBlank(protocol)) {
+                throw new URLBuilderException("Protocol of service URL is not available");
```
Comment (janakamarasena) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:442]: Add a fullstop at the end. Check other places as well.

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java [hunk 13]
```diff
@@ -199,7 +203,7 @@ public void testAddFragmentParameter() {
     public void testAddFragmentParameters() {
 
         String fragment = null;
-
+        when(CarbonUtils.getManagementTransport()).thenReturn("https");
```
Comment (janakamarasena) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java:206]: get "https" to a constant

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ServiceURL.java (hunk 11 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ServiceURL.java:40]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3111/commits/beb53f7154b1bbf61aeb6b1130af4fc849b4b681

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java (hunk 12 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:442]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3111/commits/beb53f7154b1bbf61aeb6b1130af4fc849b4b681

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java (hunk 13 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilderTest.java:206]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3111/commits/beb53f7154b1bbf61aeb6b1130af4fc849b4b681

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ServiceURL.java (hunk 11 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ServiceURL.java:40]: The framework version with this interface is not yet available in the released product.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/repository/conf/identity/identity.xml [hunk 14]
```diff
@@ -71,6 +71,9 @@
         <TrustManagerType>SunX509</TrustManagerType>
     </Security>
 
+    <!--This configuration is to resolve the internal service URL for internal API calls-->
```
Comment (Abilashini) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/repository/conf/identity/identity.xml:74]: End with a full stop

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java [hunk 15]
```diff
@@ -79,15 +80,16 @@ public ServiceURLBuilder addPath(String... paths) {
...
+        String internalHostName = fetchInternalHostName();
         int port = fetchPort();
         String tenantDomain = resolveTenantDomain();
         String proxyContextPath = ServerConfiguration.getInstance().getFirstProperty(PROXY_CONTEXT_PATH);
         String resolvedFragment = buildFragment(fragment, fragmentParams);
         String urlPath = getResolvedUrlPath(tenantDomain);
 
-        return new ServiceURLImpl(protocol, hostName, port, tenantDomain, proxyContextPath, urlPath, parameters,
-                resolvedFragment);
+        return new ServiceURLImpl(protocol, proxyHostName, internalHostName, port, tenantDomain, proxyContextPath, urlPath,
```
Comment (Abilashini) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:91]: Exceeds 120 characters

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml [hunk 16]
```diff
@@ -73,6 +73,9 @@
         <UserStorePasswordEncryption>InternalKeyStore</UserStorePasswordEncryption>
     </Security>
 
+    <!-- This configuration is to resolve the internal service URL for internal API calls -->
```
Comment (Abilashini) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml:76]: full stop

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2 [hunk 17]
```diff
@@ -70,6 +70,9 @@
         <TrustManagerType>{{key_mgt.trust_manager_type}}</TrustManagerType>
     </Security>
 
+    <!-- This configuration is to resolve the internal service URL for internal API calls -->
```
Comment (Abilashini) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2:73]: full stop

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/repository/conf/identity/identity.xml (hunk 14 above)
Comment (GANGANI) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/repository/conf/identity/identity.xml:74]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3111/commits/cb3d27b5f55b63844b8f96b218b5eb5e95124c8e

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java (hunk 15 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:91]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3111/commits/cb3d27b5f55b63844b8f96b218b5eb5e95124c8e

Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:91]: This was already fixed

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml (hunk 16 above)
Comment (GANGANI) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml:76]: Fixed https://github.com/wso2/carbon-identity-framework/pull/3111/commits/cb3d27b5f55b63844b8f96b218b5eb5e95124c8e

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java (hunk 15 above)
Comment (GANGANI) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java:91]: This has already fixed

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2 (hunk 17 above)
Comment (GANGANI) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2:73]: Fixed cb3d27b

**Pull request merged**